File: src/main/java/me/oriharel/machinery/data/PlayerMachinePersistentDataType.java
package me.oriharel.machinery.data;

import com.google.gson.Gson;
import me.oriharel.machinery.machine.MachineFactory;
import me.oriharel.machinery.machine.PlayerMachine;
import me.oriharel.machinery.utilities.Utils;
import org.bukkit.persistence.PersistentDataAdapterContext;
import org.bukkit.persistence.PersistentDataType;

public class PlayerMachinePersistentDataType implements PersistentDataType<String, PlayerMachine> {

    private Gson gson;

    public PlayerMachinePersistentDataType(MachineFactory factory) {
        this.gson = Utils.getGsonSerializationBuilderInstance(PlayerMachine.class, factory);
    }

    @Override
    public Class<String> getPrimitiveType() {
        return String.class;
    }

    @Override
    public Class<PlayerMachine> getComplexType() {
        return PlayerMachine.class;
    }

    @Override
    public String toPrimitive(PlayerMachine playerMachine, PersistentDataAdapterContext persistentDataAdapterContext) {
        return gson.toJson(playerMachine, PlayerMachine.class);
    }

    @Override
    public PlayerMachine fromPrimitive(String string, PersistentDataAdapterContext persistentDataAdapterContext) {
        return gson.fromJson(string, PlayerMachine.class);
    }
}

File: src/main/kotlin/me/oriharel/machinery/data/PlayerMachinePersistentDataType.kt
package me.oriharel.machinery.data

import com.google.gson.Gson
import me.oriharel.machinery.machines.MachineFactory
import me.oriharel.machinery.machines.machine.PlayerMachine
import me.oriharel.machinery.utilities.Utils
import org.bukkit.persistence.PersistentDataAdapterContext
import org.bukkit.persistence.PersistentDataType

class PlayerMachinePersistentDataType(factory: MachineFactory?) : PersistentDataType<String, PlayerMachine> {
    private val gson: Gson?
    override fun getPrimitiveType(): Class<String> {
        return String::class.java
    }

    override fun getComplexType(): Class<PlayerMachine> {
        return PlayerMachine::class.java
    }

    override fun toPrimitive(playerMachine: PlayerMachine, persistentDataAdapterContext: PersistentDataAdapterContext): String {
        return gson!!.toJson(playerMachine, PlayerMachine::class.java)
    }

    override fun fromPrimitive(string: String, persistentDataAdapterContext: PersistentDataAdapterContext): PlayerMachine {
        return gson!!.fromJson(string, PlayerMachine::class.java)
    }

    init {
        gson = Utils.getGsonSerializationBuilderInstance(PlayerMachine::class.java, factory)
    }
}